test/unit-test/test_u2f_standard.c

```diff
@@ -0,0 +1,399 @@
+
+    return ((resp.ctr[0] << 24) + (resp.ctr[1] << 16) + (resp.ctr[2] << 8) + (resp.ctr[3]));
+}
+
+static void check_Compilation(void)
+{
+    // Couple of sanity checks.
+    CHECK_EQ(sizeof(U2F_EC_POINT), (size_t)65);
+    CHECK_EQ(sizeof(U2F_REGISTER_REQ), (size_t)64);
+}
+
+static void run_tests(void)
+{
+    // Start of tests
+    //
+    device = U2Fob_create();
+
+    if (U2Fob_open(device) == 0) {
+        CHECK_EQ(0, U2Fob_init(device));
+        PASS(check_Compilation());
+        PASS(test_Version());
+        PASS(test_UnknownINS());
+        PASS(test_WrongLength_U2F_VERSION());
+        PASS(test_WrongLength_U2F_REGISTER());
+        PASS(test_BadCLA());
+
+        // Fob with button should need touch.
+        if (U2Fob_liveDeviceTesting() && arg_hasButton) {
+            // Timeout
+            PRINT_MESSAGE("PRESS abort or WAIT for device to timeout.\n");
+            PASS(test_Enroll(0x6985, 1));
+        }
+        WaitForUserPresence(device, arg_hasButton);
+        PASS(test_Enroll(0x9000, 1));
+
+        // Fob with button should have consumed touch.
+        if (U2Fob_liveDeviceTesting() && arg_hasButton) {
+            // Timeout
+            PRINT_MESSAGE("Press ABORT or wait for device to timeout.\n");
+            PASS(test_Sign(0x6985, false));
+        }
+
+        // Sign with check only should not produce signature.
+        PASS(test_Sign(0x6985, true));
+
+        // Sign with wrong key handle.
+        regRsp.keyHandleCertSig[0] ^= 0x55;
+        PASS(test_Sign(0x6a80, false));
+        regRsp.keyHandleCertSig[0] ^= 0x55;
+
+        // Sign with wrong app id.
+        regReq.appId[0] ^= 0xaa;
+        PASS(test_Sign(0x6a80, false));
+        regReq.appId[0] ^= 0xaa;
+
+        // Sign with check only should not produce signature.
+        WaitForUserPresence(device, arg_hasButton);
+        PASS(test_Sign(0x6985, true));
+
+        uint32_t ctr1;
+        PASS(ctr1 = test_Sign(0x9000, false)); // < fails
+        if (U2Fob_liveDeviceTesting()) {
+            // Timeout
+            PRINT_MESSAGE("Press ABORT or wait for device to timeout.\n");
+            PASS(test_Sign(0x6985, false));
+        }
+
+        WaitForUserPresence(device, arg_hasButton);
+
+        uint32_t ctr2;
+        PASS(ctr2 = test_Sign(0x9000, false));
+
+        // Ctr should have incremented by 1.
+        CHECK_EQ(ctr2, ctr1 + 1);
+
+    } else {
+        PRINT_MESSAGE("\n\nNot testing HID API. A device is not connected.\n\n");
+        return;
+    }
+
+    U2Fob_destroy(device);
+    return;
+}
+
+uint32_t __stack_chk_guard = 0;
+
+extern void __attribute__((noreturn)) __stack_chk_fail(void);
+void __attribute__((noreturn)) __stack_chk_fail(void)
+{
+    PRINT_MESSAGE("\n\nError: stack smashing detected!\n\n");
+    abort();
+}
+
+int main(void)
+{
+    // srand((unsigned int) time(NULL));
+
+    //// Test the C code API
+    // U2Fob_testLiveDevice(0);
+    //__stack_chk_guard = random_uint32(0);
```

Comment:
either set the guard or remove the stack_* stuff above.

is the stack protector even enabled in unit tests?

Reply:
not used I think.